src/StaticWebAssetsSdk/Tasks/ResolveCompressedAssets.cs

```diff
@@ -0,0 +1,309 @@
+                {
+                    existingFormats = new();
+                    existingCompressionFormatsByAssetItemSpec.Add(itemSpec, existingFormats);
+                }
+
+                var format = configuration.Format;
+                if (existingFormats.Contains(format))
+                {
+                    Log.LogMessage(
+                        "Ignoring asset '{0}' because it was already resolved with format '{1}'.",
+                        itemSpec,
+                        format);
+                    continue;
+                }
+
+                if (TryCreateCompressedAsset(asset, targetDirectory, configuration.Format, out var compressedAsset))
+                {
+                    assetsToCompress.Add(compressedAsset);
+                    existingFormats.Add(format);
+
+                    Log.LogMessage(
+                        "Created compressed asset '{0}'.",
+                        compressedAsset.ItemSpec);
+                }
+                else
+                {
+                    Log.LogError(
+                        "Could not create compressed asset for original asset '{0}'.",
+                        itemSpec);
+                }
+            }
+        }
+
+        AssetsToCompress = assetsToCompress.ToArray();
+
+        return !Log.HasLoggedErrors;
+    }
+
+    private static bool IsCompressedAsset(ITaskItem asset)
+        => string.Equals("Content-Encoding", asset.GetMetadata("AssetTraitName"));
+
+    private bool TryCreateCompressedAsset(ITaskItem asset, string targetDirectory, string format, out TaskItem result)
+    {
+        result = null;
+
+        string fileExtension;
+        string assetTraitValue;
+
+        if (CompressionConfiguration.CompressionFormat.IsGzip(format))
+        {
+            fileExtension = ".gz";
+            assetTraitValue = "gzip";
+        }
+        else if (CompressionConfiguration.CompressionFormat.IsBrotli(format))
+        {
+            fileExtension = ".br";
+            assetTraitValue = "br";
+        }
+        else
+        {
+            Log.LogError(
+                "Unknown compression format '{0}' for '{1}'.",
+                format,
+                asset.ItemSpec);
+            return false;
+        }
+
+        var originalItemSpec = asset.GetMetadata("OriginalItemSpec");
+        var relativePath = asset.GetMetadata("RelativePath");
+
+        var fileName = FileHasher.GetFileHash(originalItemSpec) + fileExtension;
+        var outputRelativePath = Path.Combine(targetDirectory, fileName);
+
+        result = new TaskItem(outputRelativePath, asset.CloneCustomMetadata());
+
+        result.SetMetadata("RelativePath", relativePath + fileExtension);
+        result.SetMetadata("OriginalItemSpec", asset.ItemSpec);
+        result.SetMetadata("RelatedAsset", asset.ItemSpec);
+        result.SetMetadata("RelatedAssetOriginalItemSpec", originalItemSpec);
```

Comment:
I see now, it's because we are putting everything in the same result.

It's ok, but when you add this to the StaticWebAsset ItemGroup use the RemoveMetadata to get rid of this property.